db/background_mgr.go

```diff
@@ -290,43 +586,103 @@ func (t *TombstoneCompactionManager) ResetStatus() {
 type AttachmentCompactionManager struct {
 	MarkedAttachments base.AtomicInt
 	PurgedAttachments base.AtomicInt
+	CompactID         string
+	Phase             string
 	lock              sync.Mutex
 }
 
 var _ BackgroundManagerProcessI = &AttachmentCompactionManager{}
 
-func NewAttachmentCompactionManager() *BackgroundManager {
+func NewAttachmentCompactionManager(bucket base.Bucket) *BackgroundManager {
 	return &BackgroundManager{
-		Process:    &AttachmentCompactionManager{},
+		Process: &AttachmentCompactionManager{},
+		clusterAwareOptions: &ClusterAwareBackgroundManagerOptions{
+			bucket:        bucket,
+			processSuffix: "compact",
+		},
 		terminator: make(chan struct{}, 1),
 	}
 }
 
-func (a *AttachmentCompactionManager) Run(options map[string]interface{}, terminator chan struct{}) error {
-	database := options["database"].(*Database)
+func (a *AttachmentCompactionManager) Init(clusterStatus []byte) error {
+	if clusterStatus != nil {
+		var attachmentResponseStatus AttachmentManagerResponse
+		err := base.JSONUnmarshal(clusterStatus, &attachmentResponseStatus)
+
+		// If the previous run completed, or there was an error during unmarshalling the status we will start the
+		// process from scratch with a new compaction ID. Otherwise, we should resume with the compact ID and phase
+		// specified in the doc.
+		if attachmentResponseStatus.State == BackgroundProcessStateCompleted || err != nil {
+			uniqueUUID, err := uuid.NewRandom()
+			if err != nil {
+				return err
+			}
+
+			a.CompactID = uniqueUUID.String()
+		} else {
+			a.CompactID = attachmentResponseStatus.CompactID
+			a.Phase = attachmentResponseStatus.Phase
+		}
+
+		return nil
 
-	uniqueUUID, err := uuid.NewRandom()
-	if err != nil {
-		return err
 	}
 
-	_, err = Mark(database, uniqueUUID.String(), terminator, &a.MarkedAttachments)
+	uniqueUUID, err := uuid.NewRandom()
 	if err != nil {
 		return err
 	}
 
-	_, err = Sweep(database, uniqueUUID.String(), terminator, &a.PurgedAttachments)
-	if err != nil {
-		return err
+	a.CompactID = uniqueUUID.String()
+
+	return nil
+}
+
+func (a *AttachmentCompactionManager) Run(options map[string]interface{}, terminator chan struct{}) error {
+	database := options["database"].(*Database)
+
+	// Need to check the current phase in the event we are resuming - No need to run mark again if we got as far as
+	// cleanup last time...
+	switch a.Phase {
+	case "mark", "":
+		a.SetPhase("mark")
+		_, err := Mark(database, a.CompactID, terminator, &a.MarkedAttachments)
+		if err != nil {
+			return err
+		}
+		fallthrough
+	case "sweep":
+		a.SetPhase("sweep")
```

Comment:
Do we need to force status persistence when we SetPhase, in case sweep does some work then fails before the next status update? (same question for cleanup below)

Reply:
Yeah. As discussed added callback to `Run` which will allow the process to force cluster status persistence